### PR TITLE
#684 JUnit 5 Extension

### PR DESCRIPTION
**Pull request by Patouche**
Hello,

I just open a draft PR to fix the #684 feature request.

For now, with this extension, we have a new annotation `@Wiremock` which will allow us to have multiple style of testing class :

```java
@ExtendWith(WireMockExtension.class)
class SampleTests {
    
    // Field injection
    @Wiremock
    private final WireMockServer server;

    private WireMockServer server2;
    
     // Constructor injection 
    SampleTests(@Wiremock(port = 8081) WireMockServer server2) {
        this.server2 = server2;
    }
  
    @Test
    void shouldSucceed() {
        testClient.get("http://localhost:8081/endpoints"));
    }
   
   // Parameter injection
   @Test
    void otherSample(@Wiremock(customizers = SampleCustomizer.class) WireMockServer server3) {
        testClient.get(String.format("http://localhost:%d/endpoints", server3.port());
    }
  
}
```

Currently, this extensions already support :
* constructor injection
* parameter injection
* field injection

We can also specified a factory to create server options, have multiple servers and factorize server responses using `Customizer`

Still, I have a question about a case that may be a nice feature. With JUnit 5, we can have nested class. What do you think about moving the customizer into a dedicated annotation to be able to factorize response between multiple servers :

```java
class SampleTest {

    @Wiremock
    private final WireMockServer server1;

    @Wiremock(name = "my-server")
    private final WireMockServer server2;

    @Nested
    @Customize(name = "server1", value = { SampleCustomizer1.class })
    @Customize(name = "my-server", value = { SampleCustomizer2class })
    class FirstNestedTest {

        @Test
        @Customize(name = "my-server", value = { SampleCustomizer3.class })
        void shouldDoSomething() {
            // ....
        }
        
    }

    @Nested
    @Customize(name = "server1", value = { SampleCustomizer3.class })
    class SecondNestedTest {
        
    }

}
```

### CODE REVIEW

**Comment by tomakehurst** [comment]
Thank you for contributing this. I'm a bit snowed-under at the moment, but I'll try and take a look tomorrow and provide some feedback.

**Comment by Patouche** [comment]
Yes, no problem. Take your time. There is no constraint no my side.

**Review by tomakehurst — CHANGES_REQUESTED** [review]
I've done a quick pass over what you've done so far. I like the ability to choose multiple means of injection.

Please see my inline comments for a few things I think could be improved in.

File: build.gradle [hunk 1]
```diff
@@ -94,6 +94,10 @@ dependencies {
     compile 'commons-fileupload:commons-fileupload:1.4'
 
     testCompile "junit:junit:4.13"
+    testImplementation("org.junit.jupiter:junit-jupiter")
```
[the remark this answers is not in the document]
Comment (Patouche) [reply build.gradle]: Yes, 

With this syntax, we cannot use bom import. Still, you're right. It's probably better to keep the same syntax everywhere.

I update the commit to cleanup the `build.gradle` file.

File: build.gradle [hunk 2]
```diff
@@ -106,6 +110,7 @@ dependencies {
         exclude group: "org.hamcrest", module: "hamcrest-library"
     }
     testCompile 'org.mockito:mockito-core:2.28.2'
```
[the remark this answers is not in the document]
Comment (Patouche) [reply build.gradle:111]: Yes, sorry, my mistake.

It was only to take a look at another mock library to get inspiration from what they're doing. In fact, Mockito is maybe not be the best for it. I just forgot to delete these lines

**Comment by tomakehurst** [comment]
I've taken this and re-worked the design (quite) a bit.

It's now all pushed on a branch and you can find the docs here: https://github.com/tomakehurst/wiremock/blob/junit-jupiter/docs-v2/_docs/junit-jupiter.md

Feedback appreciated on whether this will do everything you need and is intuitive to use. If anyone needs a beta build to try out please let me know.

**Comment by tomakehurst** [comment]
Acceptance tests are here if you'd like to see it all put together: https://github.com/tomakehurst/wiremock/tree/junit-jupiter/src/test/java/com/github/tomakehurst/wiremock/junit5

**Comment by rieckpil** [comment]
@tomakehurst this looks great 🚀  I'd be happy to test a beta build

**Comment by rieckpil** [comment]
It works excellent - I gave it a shot for a Spring Boot project and used both the declarative and programmatic approach 🚀 I'm looking forward to removing some boilerplate code in many projects 👍

**Comment by timtebeek** [comment]
Just want to say the `WireMockExtension` as seen in 2.31.0 works perfectly for our use cases as well, glad to see this added.

One question I have though is that we are extending `WireMockExtension` in classes that each replicate external services that we integrate with. That way we can offer more high level methods to stub requests for these services as separately registered extensions. I did not see this pattern mentioned in the docs, but it appears to work well. Is extending `WireMockExtension` (un)officially supported?

We wouldn't want to paint ourselves into a corner only to find this option dropped in later versions for instance.

**Comment by tomakehurst** [comment]
This is a very good question @timtebeek. Since the JUnit 4 rule supports extension I think it's fair that the Jupiter extension should too.

I'll add some  lifecycle template methods and documentation in the next release, and give some thought to guaranteeing no breaking changes.

**Comment by timtebeek** [comment]
Awesome thanks! To be clear: we are not overriding any methods from WireMockExtension for the moment, so you're welcome to make methods such as `getRuntimeInfo()` and `baseUrl()` final until a convincing use case to loosen those restrictions comes up.

We are however calling the `protected WireMockExtension(Options options, boolean configureStaticDsl, boolean failOnUnmatchedRequests, boolean proxyMode)` constructor, which has a comment above it `// Intended to be called from the builder`. We need that constructor as we want to pass Options with a separate port per extension instance, and as a consequence of that we also pass the three booleans required. Could be worthwhile to update the comment to reflect use by subclasses.

**Comment by tomakehurst** [comment]
OK, looks like the constructor is where some thought will be required. I'd quite like to avoid an explosion of constructor overloads over time as more options are added.

**Comment by tomakehurst** [comment]
Released in 2.31.0

**Pull request closed without merging**